src/category_theory/subobject.lean

```diff
@@ -0,0 +1,296 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.over
+import category_theory.thin
+import category_theory.limits.shapes.pullbacks
+import category_theory.limits.shapes.zero
+import category_theory.isomorphism_classes
+import order.modular_lattice
+
+/-!
+## Subobjects in a category.
+
+We define `subobject X` as the (isomorphism classes of) monomorphisms into `X`.
+This is naturally a preorder.
+
+When the ambient category has pullbacks, `subobject X` has an intersection operation,
+and becomes a `semilattice_inf`.
+-/
+
+universes v u
+
+noncomputable theory
+
+namespace category_theory
+
+variables {C : Type u} [category.{v} C]
+
+/-- The category of monomorphisms into `X` is "thin", i.e. a preorder. -/
+instance monomorphisms_subsingleton {X : C} (A B : { f : over X // mono f.hom }) :
+  subsingleton (A ⟶ B) :=
+begin
+  fsplit,
+  rintros ⟨f, _, wf⟩ ⟨g, _, wg⟩,
+  dsimp at *, simp at wf wg, ext, dsimp,
+  exact (@cancel_mono _ _ _ _ _ _ B.property _ _).mp (wf.trans (wg.symm)),
+end
+
+/--
+We define the subobjects of `X` simply to be the isomorphism classes of monomorphisms into `X`.
+See https://ncatlab.org/nlab/show/subobject
+
+One could instead just take the monomorphisms directly: not much changes!
+See https://mathoverflow.net/questions/184196/concise-definition-of-subobjects
+However if we follow this route we only get a preorder, not a partial order,
+which is less convenient for describing lattice properties.
+-/
+def subobject (X : C) := isomorphism_classes.obj (Cat.of { f : over X // mono f.hom })
```

Comment:
You could alternatively define this as the `thin_skeleton` of `{ f : over X // mono f.hom }`, which automatically gives you the partial order structure below (this is what I did in the topos project)